File: Backend/demo/src/main/java/com/example/demo/Service/EmpresasAsociadasService.java
package com.example.demo.Service;

import com.example.demo.DTO.EmpresaAsociadaDTO;
import com.example.demo.Entity.EmpresasAsociadas;
import com.example.demo.Repository.EmpresasAsociadasRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class EmpresasAsociadasService {

    @Autowired
    private EmpresasAsociadasRepository empresasAsociadasRepository;

    // Crear una nueva empresa
    public EmpresasAsociadas crearEmpresa(EmpresasAsociadas empresa) {
        return empresasAsociadasRepository.crear(empresa);
    }

    // Obtener todas las empresas
    public List<EmpresaAsociadaDTO> obtenerEmpresas(int page, int size) {
        return empresasAsociadasRepository.getAll(page, size);
    }

    // Actualizar empresa
    public String actualizarEmpresa(EmpresasAsociadas empresa, Integer id) {
        return empresasAsociadasRepository.update(empresa, id);
    }

    // Eliminar empresa
    public void eliminarEmpresa(Integer id) {
        empresasAsociadasRepository.delete(id);
    }

    public List<EmpresaAsociadaDTO> obtenerEmpresasConMasFallos(int page, int size) {
        return empresasAsociadasRepository.getEmpresasConMasFallos(page, size);
    }

    // Obtener empresa por ID
    public EmpresaAsociadaDTO obtenerEmpresaPorId(Integer id) {
        return empresasAsociadasRepository.getById(id);
    }

    // Obtener empresa por nombre
    public EmpresaAsociadaDTO obtenerEmpresaPorNombre(String nombre) {
        return empresasAsociadasRepository.getByName(nombre);
    }

}

File: Frontend/src/services/EmpresasAsociadas.service.js
import httpClient from "@/http-common";

// Crear una nueva empresa
const createEmpresa = (empresa, token) => {
  return httpClient.post('/empresas', empresa, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Obtener todas las empresas
const getEmpresas = (token) => {
  return httpClient.get('/empresas', {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Actualizar una empresa por ID
const updateEmpresa = (id, empresa, token) => {
  return httpClient.put(`/empresas/${id}`, empresa, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Eliminar una empresa por ID
const deleteEmpresa = (id, token) => {
  return httpClient.delete(`/empresas/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

// Obtener las empresas con más fallos
const getEmpresasConMasFallos = (token) => {
  return httpClient.get('/empresas/mas-fallos', {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
};

export default {
  createEmpresa,
  getEmpresas,
  updateEmpresa,
  deleteEmpresa,
  getEmpresasConMasFallos,
};
